include/aws/common/message_bus.h

```diff
@@ -0,0 +1,78 @@
+ * A copy of the License is located at
+ *
+ *  http://aws.amazon.com/apache2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+#include <aws/common/common.h>
+#include <aws/common/linked_list.h>
+
+/*
+ * A message bus is a mapping of integer message types -> listeners/callbacks.
+ * A listener can listen to a single message, or to all messages on a bus
+ * Message types must be consecutive integers, starting at 1. 0 is reserved
+ * for broadcast to all handlers.
+ * Handlers will be sent a message of type 0, with a NULL msg_data/user_data
+ * when it is time to clean any state up.
+ * Listeners are owned by the subscriber, and are no longer referenced by the bus
+ * once unsubscribed.
+ * Message delivery happens in a separate thread from sending, so listeners are
+ * responsible for their own thread safety
+ */
+struct aws_message_bus;
+
+/* Signature for listener callbacks */
+typedef void(aws_message_bus_listener_fn)(uintptr_t msg_type, void *msg_data, void *user_data);
+
+enum aws_message_bus_policy {
+    /* Message delivery is queued */
+    AWS_BUS_ASYNC,
+    /* Message delivery is immediate */
+    AWS_BUS_SYNC,
+};
+
+/* Creates and returns a new message bus */
+struct aws_message_bus *aws_message_bus_new(
+    struct aws_allocator *allocator, size_t buffer_size, enum aws_message_bus_policy policy);
+
+/* Cleans up a message bus by notifying all listeners and releasing the bus memory */
+void aws_message_bus_destroy(struct aws_message_bus *bus);
+
+/* Subscribes a listener to a message type */
+int aws_message_bus_subscribe(struct aws_message_bus *bus, uintptr_t address, aws_message_bus_listener_fn *listener, void *user_data);
+
+/* Unsubscribes a listener from a specific message */
+int aws_message_bus_unsubscribe(struct aws_message_bus *bus, uintptr_t address, aws_message_bus_listener_fn *listener, void *user_data);
+
+/*
+ * Allocates a new message to be sent. The message bus owns this memory, and is responsible
+ * for freeing it when message delivery is complete. If an error occurs before the message
+ * is sent, it may be freed with aws_message_bus_destroy_message()
+ */
+int aws_message_bus_new_message(struct aws_message_bus *bus, size_t size, struct aws_byte_buf **msg);
+
+/*
+ * Frees a message buffer allocated by aws_message_bus_new_message. Only necessary if the
+ * message is never sent
+ */
+void aws_message_bus_destroy_message(struct aws_message_bus *bus, struct aws_byte_buf *msg);
+
+/*
+ * Sends a message to any listeners. msg_data should have been allocated by aws_message_bus_new_message
+ * and will be cleaned up after the message is delivered
+ */
+int aws_message_bus_send(struct aws_message_bus *bus, uintptr_t address, struct aws_byte_buf *msg);
```

Comment:
We're forcing messages to be linearly laid out data. That's going to put restrictions on the user. Like, if you need to send a struct that contains a a string and an unrelated int, you either store the string in a fixed array, or you invent your own serialization format where you prefix the string with length.

Should we do like hashtable and let users set a destructor? So that any arbitrary struct/class/object/string can be passed?

I see how the setup you have here *is* ideal for sending *just* a string, it allows each message to have its own length, and stores them all in linear memory.

Reply:
We could do it that way. The main reason I didn't is I'm concerned about the cost of allocating those messages at high frequency. On the other hand, I suppose that's something we can optimize on a case by case basis?